src/backend/distributed/utils/node_metadata.c

```diff
@@ -674,21 +674,15 @@ AddNodeMetadata(char *nodeName, int32 nodePort, int32 groupId, char *nodeRack,
 	*nodeAlreadyExists = false;
 
 	/* acquire a lock so that no one can do this concurrently */
-	pgDistNode = heap_open(DistNodeRelationId(), AccessExclusiveLock);
+	/* specifically, pg_dist_node_trigger_func also takes out a ShareRowExclusiveLock */
+	pgDistNode = heap_open(DistNodeRelationId(), ShareRowExclusiveLock);
 
 	/* check if the node already exists in the cluster */
 	workerNode = FindWorkerNode(nodeName, nodePort);
 	if (workerNode != NULL)
 	{
-		/* fill return data and return */
-		returnData = GenerateNodeTuple(workerNode);
-
-		/* close the heap */
-		heap_close(pgDistNode, AccessExclusiveLock);
-
 		*nodeAlreadyExists = true;
-
-		PG_RETURN_DATUM(returnData);
+		goto finish;
```

Comment:
goto is the best way to get there ;)

maybe just use `LockRelationOid(DistNodeRelationId(), ShareRowExclusiveLock)` and avoid the `heap_close` (and the `goto`).